### ISSUE TITLE
Explanation observation_and_action_constraint_splitter

### ISSUE THREAD

**Issue by FR8803** [tensorflow/agents#550]
Hey guys, 
I'm currently trying to add an observation_and_action_constraint_splitter to my DQN network and don't really understand how to use it. First of all I would highly appreciate if you had any code examples you could share with me so I can try to understand its logic. 

I wonder whether the general idea of the splitter is to return both your observation array and the legal moves as a separate array, something like this:

def observation_and_action_constraint_splitter(observation):
...
	return observation, legal_moves

Are the legal moves then already a part of the observation being used as an input or not? What is then the improvement being achieved by introducing an observation_and_action_constraint_splitter as compared to the situation with no splitter and the legal moves as a part of the observation?

Thanks a lot for your answers!

**Comment by egonina** [comment]
Thanks for reaching out! I dug around a bit and I agree it is not very clear. CC @ebrevdo who is more familiar with the `observation_and_action_constraint_splitter`. Eugene, can you please advise? I agree we should add better documentation about this callable

**Comment by bartokg** [comment]
The most common setup with the splitter is that the original observation is a 2-tuple of (observation, legal_moves_mask) and the splitter is just lambda x: (x[0], x[1]). Then you make sure that the observations you generate for the policy are in this structure. I can show you a code snippet from the Bandit library of  TF-Agents.
In this [test](https://github.com/tensorflow/agents/blob/master/tf_agents/bandits/policies/linear_bandit_policy_test.py#L268) you see how to set up a policy that makes use of this splitter logic. It works just the same with agents.

To explain the setup a little bit: we went for this split so that if there is some observation-independent process that governs the available actions/moves, we can model that. With this decoupling we can make sure most of the agents can be fitted to this setting.

**Comment by FR8803** [comment]
Thanks @bartokg  for your quick reply! So am I understanding you right that with the splitter, you are basically giving the agent an observation plus a tuple with legal actions (which can't be observed by the agent)? What happens if the agent chooses an illegal action (not part of the legal action mask)? Is there some way of "telling" him the action was illegal? 

Basically I'm looking for a way to teach / tell the reinforcement learning agent which actions are legal and which are not. Is the observation_and_action_constraint_splitter the right way of doing so?

**Comment by ebrevdo** [comment]
The agents that respect the splitter will take the output of your neural network, and mask out the logits associated with invalid actions.  That way when the agent action is called, it is sure to never take the wrong action.

This does *not* teach the neural network that bad actions are actually bad.  For that you need a more robust/powerful algorithm than what is available in TF-Agents today, *or* an environment that can handle bad actions by sending back a very negative reward (and thus gradually teaching the agent that some actions are bad).  Since you already are providing the mask *prior* to the action, you don't need to teach it much; you can just force it to not take bad actions.  That's what the mask/splitter are for.
